mathesar/utils/users.py

```diff
@@ -0,0 +1,62 @@
+from django.db import transaction
+from django.db.models import F
+
+from mathesar.models import User
+
+
+def get_user(user_id):
+    return User.objects.get(id=user_id)
+
+
+def list_users():
+    return User.objects.all()
+
+
+@transaction.atomic
+def add_user(user_def):
+    user = User.objects.create(
+        username=user_def["username"],
+        is_superuser=user_def["is_superuser"],
+        email=user_def.get("email", ""),
+        full_name=user_def.get("full_name", ""),
+        short_name=user_def.get("short_name", ""),
+        display_language=user_def.get("display_language", "en"),
+        password_change_needed=True
+    )
+    user.set_password(user_def["password"])
+    user.save()
+    return user
+
+
+def update_user_info(user_id, user_info, requesting_user):
+    if not requesting_user.is_superuser:
+        user_info.pop("is_superuser", None)
+    User.objects.filter(id=user_id).update(
```

Comment:
I'm suspicious of having an auth-related step (making sure a non-superuser can't elevate their privileges) in this function, rather than the RPC layer function. Did you consider putting it there? We don't want to have to look in multiple files to validate privilege logic in this case, unless there's a really good reason.

Also, it seems like `User.objects.get` be better here. Why did you go with the `filter` function? Specifically, I think we'd want to throw an Exception if there's no matching object. Maybe I'm missing something.

Finally, can't you just explode the `user_info` dict for this update call? E.g.,

```
User.objects.get(id=user_id).update(**user_info)
```

Reply:
> I'm suspicious of having an auth-related step (making sure a non-superuser can't elevate their privileges) in this function, rather than the RPC layer function. Did you consider putting it there?

I did consider putting it there but thought that it would be better to add all the logic related to updating user in a single function instead of splitting it, this way we won't forget to check whether the requesting_user is a superuser, if we decided to use the `update_user_info` function elsewhere. But oh well..., we could also use the RPC functions directly instead. I've updated it as per your request.

> Also, it seems like User.objects.get be better here. Why did you go with the filter function? Specifically, I think we'd want to throw an Exception if there's no matching object. Maybe I'm missing something.

Django advises against doing this: https://docs.djangoproject.com/en/4.2/ref/models/querysets/#django.db.models.query.QuerySet.update
We _could_ raise an exception based on the number of rows changed on the db that is returned by `update()` though.

> Finally, can't you just explode the `user_info` dict for this update call? E.g.,
> 
> ```
> User.objects.get(id=user_id).update(**user_info)
> ```

We can't use `update` on a User object or any object we receive with `get`, we can only use it with `filter`. Yep, I know... kinda stupid. But I think its because `update` performs a SQL update instead of loading the result in memory first, contrary to the way `get` works.